src/coprocessor/dag/batch_executor/index_scan_executor.rs

```diff
@@ -0,0 +1,214 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use cop_datatype::EvalType;
+use kvproto::coprocessor::KeyRange;
+use tipb::expression::FieldType;
+use tipb::schema::ColumnInfo;
+
+use crate::storage::Store;
+
+use super::interface::*;
+use crate::coprocessor::codec::batch::{LazyBatchColumn, LazyBatchColumnVec};
+use crate::coprocessor::dag::expr::{EvalConfig, EvalContext};
+use crate::coprocessor::dag::Scanner;
+use crate::coprocessor::{Error, Result};
+
+pub struct BatchIndexScanExecutor<S: Store>(
+    super::scan_executor::ScanExecutor<
+        S,
+        IndexScanExecutorImpl,
+        super::ranges_iter::PointRangeConditional,
+    >,
+);
+
+impl<S: Store> BatchIndexScanExecutor<S> {
+    pub fn new(
+        store: S,
+        config: Arc<EvalConfig>,
+        columns_info: Vec<ColumnInfo>,
+        key_ranges: Vec<KeyRange>,
+        desc: bool,
+        unique: bool,
+        // TODO: this does not mean that it is a unique index scan. What does it mean?
+    ) -> Result<Self> {
+        let mut schema = Vec::with_capacity(columns_info.len());
+        let mut columns_len_without_handle = 0;
+        let mut decode_handle = false;
+        for ci in &columns_info {
+            schema.push(super::scan_executor::field_type_from_column_info(&ci));
+            if ci.get_pk_handle() {
+                decode_handle = true;
+            } else {
+                columns_len_without_handle += 1;
+            }
+        }
+
+        let imp = IndexScanExecutorImpl {
+            context: EvalContext::new(config),
+            schema,
+            columns_len_without_handle,
+            decode_handle,
+        };
+        let wrapper = super::scan_executor::ScanExecutor::new(
+            imp,
+            store,
+            desc,
+            key_ranges,
```

Comment:
Should we check `key_ranges`?

Reply:
Would be nice to check. But let's do it in another PR because we don't check it either in current executor.

Reply:
e.. It seems we have checked it in `TableScanExecutor` and `IndexScanExecutor`